### PR TITLE
std.typecons: Fix a -dip1000 compilable issue

### PR DESCRIPTION
**Pull request by carblue**
https://dlang.org/phobos/std_typecons.html#Proxy
"template Proxy(alias a)
Creates a proxy for the value a that will forward all operations while disabling implicit conversions. The aliased item a must be an lvalue."

There is this unittest in std/typecons.d:
```
@system unittest                  <== line 6579
{
    // bug14213, using function for the payload
    static struct S
    {
        int foo() { return 12; }
        mixin Proxy!foo;          <== line 6585
    }
    static class C
    {
        int foo() { return 12; }
        mixin Proxy!foo;
    }
    S s;
    assert(s + 1 == 13);
    C c = new C();
    assert(s * 2 == 24);
}
```

I don't understanf, why a non-lvalue returning function is used/(usable at all) here as template argument, but it compiles for S and C with -dip25.
With -dip1000 it compiles for C only.
Strange.
This quasi-fix just comments-out the offending lines to be -dip1000 compilable, but as said, without understanding much of what the implications are.

https://github.com/dlang/phobos/blob/master/dip1000.mak with
aa[std.typecons]=-dip1000
Errors when running: make -f posix.mak std/typecons.test
...
```
T=`mktemp -d /tmp/.dmd-run-test.XXXXXX` &&                                                              \
  (                                                                                                     \
    ../dmd/generated/linux/release/64/dmd -od$T -conf= -I../druntime/import  -w -de -dip25 -m64 -fPIC -transition=complex -O -release -dip1000  -main -unittest generated/linux/release/64/libphobos2.a -defaultlib= -debuglib= -L-ldl -cov -run std/typecons.d ;     \
    RET=$? ; rm -rf $T ; exit $RET                                                                   \
  )
std/typecons.d(5953): Error: circular typeof definition
std/typecons.d(5953): Error: alias `std.typecons.__unittest_L6579_C9.S.Proxy!(foo).ValueType` recursive alias declaration
std/typecons.d(6585): Error: mixin `std.typecons.__unittest_L6579_C9.S.Proxy!(foo)` error instantiating
```

### CODE REVIEW

File: std/typecons.d [hunk 1]
```diff
@@ -6582,17 +6582,17 @@ mixin template Proxy(alias a)
...
     }
     static class C
     {
         int foo() { return 12; }
         mixin Proxy!foo;
     }
-    S s;
-    assert(s + 1 == 13);
+//    S s;
+//    assert(s + 1 == 13);
```
Comment (wilzbach) [inline std/typecons.d]: This removes the test for the normal code path too, so if we really decide we don't want to test this for DIP1000, I think it should be sth. like:

```
unittest
{
    static class C
...
}

version(DIP1000) {} else
unittest
{
   static struct S
}
```

Comment (WalterBright) [reply std/typecons.d]: There isn't a standard DIP1000 version (perhaps there should be), but in any case the reasons these tests are commented out should be made clear in the code.

File: std/typecons.d [hunk 2]
```diff
@@ -6829,6 +6829,18 @@ mixin template Proxy(alias a)
...
+@system unittest
+{
+    // bug14213, using function for the payload
+    static class C
+    {
+        int foo() { return 12; }
+        mixin Proxy!foo;
+    }
+    C c = new C();
+}
```
Comment (wilzbach) [inline std/typecons.d]: The duplication wouldn't have been necessary:

```d
@system unittest
{
    // bug14213, using function for the payload
    static struct S
    {
        int foo() { return 12; }
        mixin Proxy!foo;
    }
    S s;
    assert(s + 1 == 13);
    assert(s * 2 == 24);
}

version(DIP1000) {} else
@system unittest
{
    static class C
    {
        int foo() { return 12; }
        mixin Proxy!foo;
    }
    C c = new C();
}

Comment (carblue) [reply std/typecons.d]: @wilzbach Yes, I was aware of the duplicated class C: My idea was to simply delete my original version(DIP1000) addition again, once the reason for failing struct S mixin Proxy!foo is found.

Comment (wilzbach) [reply std/typecons.d]: Deleting `version(DIP1000) {} else` is even easier ;-)

**Pull request merged**